util/include/Metrics.hpp

```diff
@@ -0,0 +1,186 @@
+  Status GetStatus(const std::string& component_name,
+                   const std::string& val_name);
+  Counter GetCounter(const std::string& component_name,
+                     const std::string& val_name);
+
+ private:
+  void RegisterComponent(Component& component);
+  void UpdateValues(const std::string& name, Values&& values);
+
+  std::map<std::string, Component> components_;
+  std::mutex lock_;
+
+  friend class Component;
+};
+
+// A Gauge is a an integer value that shows the current value of something. It
+// can only be varied by directly setting and getting it.
+class Gauge {
+ public:
+  explicit Gauge(const uint64_t val) : val_(val) {}
+
+  void Set(const uint64_t val) { val_ = val; }
+  uint64_t Get() { return val_; }
+
+ private:
+  uint64_t val_;
+};
+
+// Status is a text based representation of a value. It's used for things that
+// don't have strictly numeric representations, like the current state of the
+// BFT or the last message received.
+class Status {
+ public:
+  explicit Status(const std::string& val) : val_(val) {}
+
+  void Set(const std::string& val) { val_ = val; }
+  std::string Get() { return val_; }
+
+ private:
+  std::string val_;
+};
+
+class Counter {
+ public:
+  explicit Counter(const uint64_t val) : val_(val) {}
+
+  // Increment the counter and return the value after incrementing.
+  uint64_t Inc() { return ++val_; }
+
+  uint64_t Get() { return val_; }
+
+ private:
+  uint64_t val_;
+};
+
+class Values {
+ private:
+  std::vector<Gauge> gauges_;
+  std::vector<Status> statuses_;
+  std::vector<Counter> counters_;
+
+  friend class Component;
+  friend class Aggregator;
+};
+
+// We keep the names of values in separate vecs since they remain constant for
+// the life of the program. When we update the component in the aggregator we
+// don't have to copy all the names every time. We just have to do it once
+// during initialization.
+class Names {
+ private:
+  std::vector<std::string> gauge_names_;
+  std::vector<std::string> status_names_;
+  std::vector<std::string> counter_names_;
+
+  friend class Component;
+  friend class Aggregator;
+};
+
+// A Component stores Values of different types and is updated on the local
+// thread. Components are sent to an Aggregator periodically. Components are
+// optimized for fast update access.
+class Component {
+ public:
+  // A Handle allows for fast access to the underlying value inside the
+  // component via an index operation.
+  //
+  // Note that the handle cannot live longer than the Component, and should
+  // never be used from a separate thread.
+  //
+  // In concord-bft we expect components to live for the lifetime of the program
+  // and handles to be member variables used to update values on the same
+  // thread.
+  template <typename T>
+  class Handle {
+   public:
+    Handle(std::vector<T>& values, size_t index)
+        : values_(values), index_(index) {}
+    T& Get() { return values_[index_]; }
```

Comment:
I would check an index for validity, otherwise, values_[index_] could throw an out-of-range exception.

Reply:
That can't happen in this code. All the indexes are fixed when the handle gets created. I specifically don't check for speed reasons.